File: backend/api/routes/documents.py
from __future__ import annotations

from fastapi import APIRouter, Depends, File, Query, Request, UploadFile
from sqlalchemy.orm import Session

from backend.api.deps.database import get_db_session
from backend.api.schemas.documents import (
    BatchDeleteDocumentsRequest,
    DocumentDetailData,
    UploadDocumentData,
)
from backend.api.schemas.response import success_response
from backend.app.repositories.document_repository import DocumentRepository
from backend.app.services.document_service import create_document_upload, delete_document, get_document_detail


router = APIRouter(prefix="/documents", tags=["documents"])


@router.get("")
def list_documents(
    search: str | None = Query(default=None),
    sort: str = Query(default="created_at"),
    order: str = Query(default="desc"),
    db_session: Session = Depends(get_db_session),
) -> dict:
    documents = DocumentRepository(db_session).list_documents(
        search=search,
        sort_by=sort,
        order=order,
    )
    return success_response(
        message="获取文档列表成功",
        data=[
            {
                "id": document.id,
                "name": document.name,
                "file_type": document.file_type,
                "status": document.status,
                "storage_path": document.storage_path,
                "created_at": document.created_at.isoformat(),
                "updated_at": document.updated_at.isoformat(),
            }
            for document in documents
        ],
    )


@router.post("/upload")
async def upload_document(
    request: Request,
    file: UploadFile = File(...),
    db_session: Session = Depends(get_db_session),
) -> dict:
    content = await file.read()
    document, task = create_document_upload(
        db_session,
        settings=request.app.state.settings,
        filename=file.filename,
        content=content,
    )
    return success_response(
        message="文档上传成功",
        data=UploadDocumentData(
            document_id=document.id,
            task_id=task.id,
        ).model_dump(),
    )


@router.post("/batch-upload")
async def batch_upload_documents(
    request: Request,
    files: list[UploadFile] = File(...),
    db_session: Session = Depends(get_db_session),
) -> dict:
    results = []
    for file in files:
        content = await file.read()
        document, task = create_document_upload(
            db_session,
            settings=request.app.state.settings,
            filename=file.filename,
            content=content,
        )
        results.append(
            UploadDocumentData(document_id=document.id, task_id=task.id).model_dump()
        )
    return success_response(message="批量上传成功", data=results)


@router.post("/batch-delete")
def batch_delete_documents(
    request: Request,
    payload: BatchDeleteDocumentsRequest,
    db_session: Session = Depends(get_db_session),
) -> dict:
    for document_id in payload.document_ids:
        delete_document(db_session, document_id, settings=request.app.state.settings)
    return success_response(message="批量删除成功")


@router.get("/{document_id}")
def get_document(
    document_id: str,
    db_session: Session = Depends(get_db_session),
) -> dict:
    document = get_document_detail(db_session, document_id)
    return success_response(
        message="获取文档详情成功",
        data=DocumentDetailData(
            id=document.id,
            name=document.name,
            file_type=document.file_type,
            status=document.status,
            storage_path=document.storage_path,
            created_at=document.created_at.isoformat(),
            updated_at=document.updated_at.isoformat(),
        ).model_dump(),
    )


@router.delete("/{document_id}")
def remove_document(
    request: Request,
    document_id: str,
    db_session: Session = Depends(get_db_session),
) -> dict:
    delete_document(db_session, document_id, settings=request.app.state.settings)
    return success_response(message="文档删除成功")

File: frontend/src/services/documents.ts
import { requestJson } from "@/services/http"

export interface UploadDocumentData {
  document_id: string
  task_id: string
}

export interface DocumentDetailData {
  id: string
  name: string
  file_type: string
  status: string
  storage_path: string
  created_at: string
  updated_at: string
}

export interface TaskDetailData {
  id: string
  document_id: string
  task_type: string
  status: string
  error_message: string | null
  created_at: string
  updated_at: string
}

export async function uploadDocument(file: File): Promise<UploadDocumentData> {
  const formData = new FormData()
  formData.append("file", file)

  const response = await requestJson<UploadDocumentData>("/api/documents/upload", {
    method: "POST",
    body: formData,
  })

  if (!response.data) {
    throw new Error("文档上传成功，但服务未返回文档信息。")
  }

  return response.data
}

export async function fetchDocument(documentId: string): Promise<DocumentDetailData> {
  const response = await requestJson<DocumentDetailData>(`/api/documents/${documentId}`)

  if (!response.data) {
    throw new Error("服务未返回文档详情。")
  }

  return response.data
}

export async function removeDocument(documentId: string): Promise<void> {
  await requestJson<null>(`/api/documents/${documentId}`, {
    method: "DELETE",
  })
}

export async function searchDocuments(
  search?: string,
  sort?: string,
  order?: "asc" | "desc",
): Promise<DocumentDetailData[]> {
  const queryParams = new URLSearchParams()

  if (search) {
    queryParams.append("search", search)
  }

  if (sort) {
    queryParams.append("sort", sort)
  }

  if (order) {
    queryParams.append("order", order)
  }

  const response = await requestJson<DocumentDetailData[]>(
    `/api/documents?${queryParams.toString()}`,
  )

  if (!response.data) {
    throw new Error("服务未返回文档列表。")
  }

  return response.data
}

export async function batchDeleteDocuments(documentIds: string[]): Promise<void> {
  await requestJson<null>("/api/documents/batch-delete", {
    method: "POST",
    body: JSON.stringify({ document_ids: documentIds }),
  })
}
